lib/srv/alpnproxy/kube.go

```diff
@@ -66,13 +99,92 @@ func (m *KubeMiddleware) OverwriteClientCerts(req *http.Request) ([]tls.Certific
 		return nil, trace.BadParameter("expect a TLS request")
 	}
 
+	m.certsMu.RLock()
 	cert, ok := m.certs[req.TLS.ServerName]
+	m.certsMu.RUnlock()
 	if !ok {
 		return nil, trace.NotFound("no client cert found for %v", req.TLS.ServerName)
 	}
+
+	var err error
+	cert, err = m.reissueCertIfExpired(req.Context(), cert, req.TLS.ServerName)
+	if err != nil {
+		// If we exceeded timeout, we return an error that will try to get user attention to the local proxy
+		if errors.Is(err, ErrUserInputRequired) {
+			return nil, trace.Wrap(&apierrors.StatusError{
+				ErrStatus: metav1.Status{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Status",
+						APIVersion: "v1",
+					},
+					Status:  metav1.StatusFailure,
+					Code:    http.StatusGatewayTimeout,
+					Reason:  metav1.StatusReasonTimeout,
+					Message: "Local Kube proxy requires user input to continue",
+				},
+			})
+		}
+		return nil, trace.Wrap(err)
+	}
+
 	return []tls.Certificate{cert}, nil
 }
 
+// ErrUserInputRequired returned when user's input required to relogin and/or reissue new certificate.
+var ErrUserInputRequired = errors.New("user input required")
+
+// reissueCertIfExpired checks if provided certificate has expired and reissues it if needed, otherwise it returns original certificate.
+func (m *KubeMiddleware) reissueCertIfExpired(ctx context.Context, cert tls.Certificate, serverName string) (tls.Certificate, error) {
+	x509Cert, err := utils.TLSCertLeaf(cert)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	if err := utils.VerifyCertificateExpiry(x509Cert, m.clock); err == nil {
+		return cert, nil
+	}
+
+	// If certificate has expired we try to reissue it.
+	identity, err := tlsca.FromSubject(x509Cert.Subject, x509Cert.NotAfter)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	type certResult struct {
+		cert tls.Certificate
+		err  error
+	}
+
+	certCh := make(chan certResult, 1)
+	// We start cert reissuing (with relogin if required) only if it's not running already.
+	// After that it will run until user gives required input.
+	// User requests will return error notifying about required user input while reissuing is running.
+	if m.isCertReissuingRunning.CompareAndSwap(false, true) {
```

Comment:
So the reissuing is one kube cluster at a time?

Reply:
Main idea of this is so we quickly return (after 3 seconds currently) and error to kubectl asking for user's attention. And all consecutive kubectl calls immediately return same error. And user can go to the local proxy terminal and it's ready to login, just waiting the input.
But yes, reissuing is one cluster at a time, it's on demand - when we encounter request to the cluster without valid cert.